### PR TITLE
Add back 24.05 response sending path

### PR DESCRIPTION
**Pull request by Tabrizian**
Before:
```
        "infbench_summary": {
            "Requests": 1000,
            "Requests/s": 56.781948247187536,
            "E2E Latency(s)": 17.61123087300075,
            "Avg Latency(s)": 16.54096719381794,
            "Total Input Tokens": 127000,
            "Total Output Tokens": 128000,
            "Avg. Input Tokens / Request": 127.0,
            "Avg. Output Tokens / Request": 128.0,
            "Stddev Output Tokens / Request": 0.0,
            "Input Tokens/s": 7211.307427392818,
            "Output Tokens/s": 7268.089375640005,
            "Request Issues/s": 2999.715792703272,
            "Avg TTFT(s)": 2.12225890412401,
            "Avg TPOT(s)": 0.1135331361393223
        }
```

After:

```
        "infbench_summary": {
            "Requests": 1000,
            "Requests/s": 67.5828421712087,
            "E2E Latency(s)": 14.796655007001391,
            "Avg Latency(s)": 12.20483882850504,
            "Total Input Tokens": 127000,
            "Total Output Tokens": 128000,
            "Avg. Input Tokens / Request": 127.0,
            "Avg. Output Tokens / Request": 128.0,
            "Stddev Output Tokens / Request": 0.0,
            "Input Tokens/s": 8583.020955743505,
            "Output Tokens/s": 8650.603797914713,
            "Request Issues/s": 3186.022902867275,
            "Avg TTFT(s)": 2.106685652077096,
            "Avg TPOT(s)": 0.07951301713722789
        }
```

24.05:

```
        "infbench_summary": {
            "Requests": 1000,
            "Requests/s": 66.70461264839375,
            "E2E Latency(s)": 14.991467011001077,
            "Avg Latency(s)": 13.56315037568501,
            "Total Input Tokens": 127000,
            "Total Output Tokens": 128000,
            "Avg. Input Tokens / Request": 127.0,
            "Avg. Output Tokens / Request": 128.0,
            "Stddev Output Tokens / Request": 0.0,
            "Input Tokens/s": 8471.485806346005,
            "Output Tokens/s": 8538.1904189944,
            "Request Issues/s": 2837.3298150774403,
            "Avg TTFT(s)": 2.1147114573539767,
            "Avg TPOT(s)": 0.09014518833331522
        }
```

### CODE REVIEW

File: src/pb_stub.cc [hunk 1]
```diff
@@ -784,12 +809,32 @@ Stub::ProcessReturnedResponses(
...
+    for (size_t i = 0; i < responses_size; i++) {
+      // Check the return type of execute function.
+      InferRequest* infer_request = py_requests[i].cast<InferRequest*>();
+      InferResponse* infer_response = py_responses[i].cast<InferResponse*>();
+      infer_response->PruneOutputTensors(
+          infer_request->RequestedOutputNames());
+      ProcessResponse(infer_response);
+      responses_shm_handle[i] = infer_response->ShmHandle();
+    }
+    response_batch_shm_ptr->batch_size = requests_size;
```
Comment (kthui) [inline src/pb_stub.cc:837]:
```suggestion
  for (size_t i = 0; i < responses_size; i++) {
    // Check the return type of execute function.
    InferRequest* infer_request = py_requests[i].cast<InferRequest*>();
    InferResponse* infer_response = py_responses[i].cast<InferResponse*>();
    infer_response->PruneOutputTensors(
        infer_request->RequestedOutputNames());
    ProcessResponse(infer_response);
    responses_shm_handle[i] = infer_response->ShmHandle();
  }
  response_batch_shm_ptr->batch_size = requests_size;
```
Looks like there is an extra indentation to the right?

File: src/pb_stub.cc [hunk 2]
```diff
@@ -784,12 +809,32 @@ Stub::ProcessReturnedResponses(
...
       }
+
       std::shared_ptr<InferResponse> response =
           py_responses[i].cast<std::shared_ptr<InferResponse>>();
-      request->GetResponseSender()->Send(
-          response, TRITONSERVER_RESPONSE_COMPLETE_FINAL);
+      request->GetResponseSender()->UpdateStateAndCounters(response, TRITONSERVER_RESPONSE_COMPLETE_FINAL);
     }
   }
+  response_batch = std::move(shm_pool_->Construct<char>(sizeof(IPCMessageShm) +
```
Comment (kthui) [inline src/pb_stub.cc:818]:
```suggestion
  }
  // Initialize the 'response_batch' and send the returned response
  response_batch = std::move(shm_pool_->Construct<char>(sizeof(IPCMessageShm) + 
```
Maybe adding a comment on what will happen next?

File: src/pb_stub.cc [hunk 3]
```diff
@@ -732,11 +730,38 @@ Stub::ProcessRequests(RequestBatch* request_batch_shm_ptr)
       request->GetResponseSender()->Close();
     }
   }
+
+  if (!response_batch) {
+      response_batch = shm_pool_->Construct<char>(sizeof(ResponseBatch) + sizeof(IPCMessageShm));
+      ResponseBatch* response_batch_shm_ptr =reinterpret_cast<ResponseBatch*>(response_batch.value().data_.get() + sizeof(IPCMessageShm));
+      response_batch_shm_ptr->batch_size = 0;
```
Comment (kthui) [inline src/pb_stub.cc:737]:
```suggestion
    response_batch = shm_pool_->Construct<char>(sizeof(ResponseBatch) + sizeof(IPCMessageShm));
    ResponseBatch* response_batch_shm_ptr =reinterpret_cast<ResponseBatch*>(response_batch.value().data_.get() + sizeof(IPCMessageShm));
    response_batch_shm_ptr->batch_size = 0;
```
Looks like there is an extra indentation to the right?

**Review by kthui — APPROVED** [review]
Some minor comments. LGTM, assuming the CI passes.

File: src/pb_stub.cc (hunk 3 above)
Comment (Tabrizian) [reply src/pb_stub.cc:737]: Fixed.

File: src/pb_stub.cc (hunk 1 above)
Comment (Tabrizian) [reply src/pb_stub.cc:837]: Good catch, fixed.

**Pull request closed without merging**